### ISSUE TITLE
Move custom layers to server's database

### ISSUE THREAD

**Issue by HarelM** [IsraelHikingMap/Site#364]
There's no reason a logged-in user should define his custom layers on every client he's using or on every PC.
This however, raises the question what to do with a layer that was sent as part of a site share?
One option is that this layer will be temporary stored as long as the session is alive as the user did not actively added this layer.
Another option is to store is much like it is stored today and add it to the database.
The concept of local storage is not right IMO for this kind of data.
@valleyofdawn, @zstadler I would appreciate your opinion.
This also reduces the functionality non-registered users will have.
I think that since this is an advanced functionality and most of the site works without registering I don't see this as a problem, we will need to notify the users on such a change though.

**Comment by zstadler** [comment]
I'm in favor of storing the private layers on the server for registered OSM users. This can also be used as a migration tool if and when https becomes mandatory. 

In order to minimize negative impact on our users I suggest the following:
- Layers added via sharing will continue to be added to the private layers. This is an important mechanism for sharing layers, as we've seen on the Facebook page. 
- When the local storage is saved at the server, it should be merged with the server side data in order to avoid loosing any private layers defined in one client and not the others.

There is a question of what layers registered users should see after logging out. If the layers are no longer stored locally, they will not see any private layers. IMO, that should not be an issue.

**Comment by HarelM** [comment]
I agree the logout scenario is not interesting.
I'm not sure I fully understand what you propose, can you elaborate on the use cases?
1. User with OSM account, no private layers, gets a share
2. User with OSM account, with private layers (on server? on local storage?), gets a share
3. User without an OSM account, no private layers, gets a share
4. User without an OSM account, with private layers (where are they stored now?), gets a share

I think the migration process should be simple, here's what I propose: 
for every layer in local storage for OSM users layers should be sent to server and a local storage flag should be set that the layers are in the server.
Every new client will do the same process so that eventually all layers will be sent to the server and all the clients will have a flag in local storage indicating the layer are on the server - this is what you describe as the merge procedure.

So once this migration is finished the only thing that is stored in local storage is the fact that the layers are on the server.
I'm not sure what to do with non logged in users - I think they should be less privileged.
In the above case, we should notify the Facebook group that in the (near?) future we plan to stop storing private layers for non-OSM users and who ever would like his private layers to be stored should go and register in OSM.

Thoughs are welcome.

**Comment by zstadler** [comment]
The change of the storage location should not change the current behavior of layers and sharing for OSM users. 

For users without an OSM account (actually, all unauthenticated sessions) there will be a transition period of several months after completing this task, where the site will continue to support private layers using the local storage. Afterwards, local storage will be ignored. 

I agree that all layers of OSM users should migrate to the server as soon as they authenticate from a given client, and that the server accumulates the layers from all clients. 

The migration from local storage should continue after the transition period! This will give another nudge for users to login in order to restore their layers, and it will enable OSM users to accumulate layers from clients regardless of the migration period

Assuming that in all the cases below, the  share has a non-standard layer:
> 1. User with OSM account, no private layers, gets a share
> 2. User with OSM account, with private layers (on server? on local storage?), gets a share

In both cases the non-standard layer is added to the user's private layers on the server.

> 3. User without an OSM account, no private layers, gets a share
> 4. User without an OSM account, with private layers (where are they stored now?), gets a share

In both cases:
- During the transition period, the non-standard layer is added to the user's private layers on local storage. 
- After the transition period, the non-standard layer will not be available after the browser window/tag is closed.

**Comment by HarelM** [comment]
I don't have the capacity to support this transition period as you described it - I suspect this will cause a code that will have a lot of bugs which will prevent other, probably more important features to be implemented.
Therefore, I think the migration period should start now by announcing it on facebook.
Current behavior will not change and it still reflects the definition of your migration period as far as I can understand, please correct me if I'm wrong.
When the implementation of this feature will be completed - Layers will be saved on the server for registered users and not saved for non-registered it will have the same affect as finishing your migration period. 
Also several months is too much, IMO. I think around a month should be enough.

**Comment by zstadler** [comment]
I humbly disagree with the cost & complexity estimation above.

Currently, [the update of base layers in local storage](https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.Web/sources/application/services/layers/LayersService.ts#L130-L131) is:
```Javascript
this.storedBaseLayers.push(layerData);
this.storedBaseLayers = this.unique(this.storedBaseLayers);
```
Where `storedBaseLayers` is defined using `@LocalStorage()`.


In the long-term, `storedBaseLayers` will be removed, and the code would use a member that is copied to the server. Something along the lines:
```Javascript
this.baseLayers.push(layerData);
this.baseLayers = this.unique(this.baseLayers);
if (this.osmUserService.isLoggedIn()) {
  this.setRemoteBaseLayers(this.baseLayers);
}
```

During the migration period the code would be have an else clause:
```Javascript
this.baseLayers.push(layerData);
this.baseLayers = this.unique(this.baseLayers);
if (this.osmUserService.isLoggedIn()) {
  this.setRemoteBaseLayers(this.baseLayers);
} else { // Remove after the migration period is over
  this.storedBaseLayers = this.baseLayers;
}
```

Similarly for the overlayLayers, and for restoring the saved layers.

**Comment by HarelM** [comment]
I'm not sure this is the only point in the code that will be affected by the migration period but I also though about it some more and the difference in storing the layers can probably be confined to a method called for example `storeBaseLayers(...)` which will do the job so the complexity might be low.
But another issue I thought about is that if we move the layers to the server it might take some time to fetch them - which might cause the code to be a lot more complex since the layerService will not be "ready" until the layers are fetched, which means we'll need to wait for this request to finish everywhere we use the layers service.
So I thought there should be both local storage and server storage and the code should be able to sync between them, what do you think?

I'm also thinking about the split and merge route feature - which makes be believe there should be two different services - `RoutesService` and `LayerService` (which uses the `RouteService` or not haven't fully thought about it).
I haven't fully thought about it but the only place these two together is in the UI which can use two services instead of one...

Again, thoughts are welcome...

**Comment by zstadler** [comment]
In my view, and the code above, the server is accessed only when the list of layers is restored or changed.
- Layer switching is done using only client side objects of the `LayersServices` class and should be at least as fast as currently. 
- A change in the layers list would send an update to the server and will not create a delay for the user.
- The time restoring the layers from server is probably negligible W.R.T. the overall time of the initialization, login, and authentication operations.

I agree that handling of routes should be moved out of `LayersService.ts` into its own `RoutesService.ts`.

**Comment by HarelM** [comment]
Server code was updated. It will work only from test site for now as most of the code is in the UI.
The following behavior was implemented:
1. Layers are store in the database and in local storage.
2. Current layers that are in the browser's local storage are added to the database once initialization is complete.
3. Active layer are still kept in local storage (I thought it made more sense).
4. When user logs off the layers are kept.
5. When user logs on the layers from the server are added to those in the local storage and all layers are saved to database - cross browser layer sharing works!!!
6. Made very little tests about custom layers that were shared, but they should be save to database too.

@zstadler @valleyofdawn
More tests are needed so please help me out...
If you don't have a backup to your layers please don't test it out since I'm not 100% sure all the code works - so at your own risk, although I made some test and they gave good results...

**Comment by valleyofdawn** [comment]
Works for me in the test site

**Comment by zstadler** [comment]
**Use case**: A user has logged-in from multiple devices/local storages, layer definitions were merged on the server, and each devices/local storage was refreshed with definitions from the server. Afterwards, a layer is deleted in one device/local storage

**What I would expect to happen**: When another devices/local storage is refreshed, the deleted layer is removed from each device/local storage.

**What actually happens**: When another device/local storage is refreshed, the deleted layer is kept, and restored to the server definition. 

As a result, deleting a layer is possible only if the deletion is done on every device/local storage sequentially, without any refresh in-between.

I believe this is because local storage is still used after merging its data to the server, and it keeps updating the server even after the server changes should override it.

Looking back on our approach, at a minimum, the local storage should capture the fact that it was already merged into the server. From that point on, the server storage is determining the layers list. 

Frankly, I'm not sure that there is any point keeping the layer configuration in local storage after the merge to the server has been done.

**Comment by HarelM** [comment]
Yes, you are right, I haven't thought about this scenario. I don't see a good reason to continue storing the layers in local storage but for some reason I kept it, probably related to non OSM users. I'll try and think of a solution that will solve both cases...

**Comment by valleyofdawn** [comment]
It may not be a bad idea to penalize non-OSM-users. People who use custom maps are advanced users and should be lured to OSM.

**Comment by zstadler** [comment]
There is no reason to penalize or coerce anyone. I want people to join OSM for we offer, such as updating the map and synchronizing private layers across devices and http vs. https.

Unlike what I wrote above, now I also don't see a reason to keep private layers layers in local storage after user login.

**Comment by HarelM** [comment]
I'll be starting the "migration" period after this check-in (and publish).
This means that whoever would like to have private layers stored (after the migration period is complete - probably around a month or two) should join OSM otherwise the layer will not be available, this is what we agreed in this thread, this also include the previous key that was used for private layers.
In any case, I'll post on Facebook to let our users know.
